collections/nemo_asr/nemo_asr/data_layer.py

```diff
@@ -413,3 +441,98 @@ def forward(self, in_x, in_x_len, in_y, in_y_len):
         out_y_len = in_y_len.repeat(self.mult)
 
         return out_x, out_x_len, out_y, out_y_len
+
+
+class TranscriptDataLayer(DataLayerNM):
+    """A simple Neural Module for loading textual transcript data.
+    The path, labels, and eos_id arguments are dataset parameters.
+
+    Args:
+        pad_id (int): Label position of padding symbol
+        batch_size (int): Size of batches to generate in data loader
+        drop_last (bool): Whether we drop last (possibly) incomplete batch.
+            Defaults to False.
+        num_workers (int): Number of processes to work on data loading (0 for
+            just main process).
+            Defaults to 0.
+        dataset_type: The type of dataset used for this datalayer.
+            Defaults to TranscriptDataset.
+    """
+
+    @staticmethod
+    def create_ports():
+        input_ports = {}
+        output_ports = {
+            'texts': NeuralType({
+                0: AxisType(BatchTag),
+                1: AxisType(TimeTag)
+            })
+        }
+        return input_ports, output_ports
+
+    def __init__(self,
+                 path,
+                 labels,
+                 eos_id,
+                 pad_id,
+                 batch_size,
+                 drop_last=False,
+                 num_workers=0,
+                 dataset_type=TranscriptDataset,
+                 **kwargs):
+        super().__init__(**kwargs)
+
+        # Set up dataset
+        dataset_params = {'path': path,
+                          'labels': labels,
+                          'eos_id': eos_id}
+
+        if isinstance(dataset_type, str):
+            dataset_type = getattr(sys.modules[__name__], dataset_type)
+        self._dataset = dataset_type(**dataset_params)
+
+        # Set up data loader
+        if self._placement == DeviceType.AllGpu:
+            sampler = torch.utils.data.distributed.DistributedSampler(
+                    self._dataset)
+        else:
+            sampler = None
+
+        # noinspection PyTypeChecker
+        self._dataloader = torch.utils.data.DataLoader(
+            dataset=self._dataset,
+            batch_size=batch_size,
+            collate_fn=partial(self._collate_fn, pad_id=pad_id, pad8=True),
+            drop_last=drop_last,
+            shuffle=sampler is None,
+            sampler=sampler,
+            num_workers=num_workers
+        )
+
+    @staticmethod
+    def _collate_fn(batch_list, pad_id, pad8=False):
+        max_len = max(len(s) for s in batch_list)
+        if pad8:
+            max_len = nemo_pad_to(max_len, 8)
+
+        texts = torch.empty(len(batch_list), max_len,
+                            dtype=torch.long)
+        texts.fill_(pad_id)
+
+        for i, s in enumerate(batch_list):
+            texts[i].narrow(0, 0, s.size(0)).copy_(s)
+
+        assert len(texts.shape) == 2
```

Comment:
we should check and raise Error instead of assert.

Reply:
@redoctopus yes, please do not use "assert" - raise an exception instead